internal/backend/runtime/omni/audit/log_file.go

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2024 Sidero Labs, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+
+package audit
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/siderolabs/gen/pair/ordered"
+
+	"github.com/siderolabs/omni/internal/pkg/pool"
+)
+
+// LogFile is a rotating log file.
+//
+//nolint:govet
+type LogFile struct {
+	dir string
+
+	mu        sync.Mutex
+	f         *os.File
+	lastWrite time.Time
+
+	pool pool.Pool[bytes.Buffer]
+}
+
+// NewLogFile creates a new rotating log file.
+func NewLogFile(dir string) *LogFile {
+	return &LogFile{
+		dir: dir,
+		pool: pool.Pool[bytes.Buffer]{
+			New: func() *bytes.Buffer {
+				return &bytes.Buffer{}
+			},
+		},
+	}
+}
+
+// Dump writes data to the log file, creating new one on demand.
+func (l *LogFile) Dump(data any) error {
+	return l.dumpAt(data, time.Time{})
+}
+
+func (l *LogFile) dumpAt(data any, at time.Time) error {
+	b := l.pool.Get()
+	defer func() { b.Reset(); l.pool.Put(b) }()
+
+	err := json.NewEncoder(b).Encode(data)
+	if err != nil {
+		return err
+	}
+
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	if at.IsZero() {
+		at = time.Now()
+	}
+
+	f, err := l.openFile(at)
+	if err != nil {
+		return err
+	}
+
+	_, err = io.Copy(f, b)
```

Comment:
if we just `Write(b.Bytes())` that would be atomic, we don't need to lock the writes.

Reply:
Yeah, but that will break `lastWrite` field access, and since `*os.File.Write` uses mutex internally I don't think it makes a lot of sense to unlock and then atomically write `lastWrite` field.

Reply:
Replaced `io.Copy` with `b.WriteTo(f)`